src/goto-symex/goto_symex.h

```diff
@@ -528,6 +530,103 @@ class goto_symext
   /// \param code: The assignment to execute
   void symex_assign(statet &state, const code_assignt &code);
 
+  /// Attempt to constant propagate side effects of the assignment (if any)
+  ///
+  /// \param state: goto symex state
+  /// \param symex_assign: object handling symbol assignments
+  /// \param lhs: lhs of the assignment
+  /// \param rhs: rhs of the assignment
+  /// \return true if the operation could be evaluated to a constant string,
+  ///   false otherwise
+  bool constant_propagate_assignment_with_side_effects(
+    statet &state,
+    symex_assignt &symex_assign,
+    const exprt &lhs,
+    const exprt &rhs);
+
+  /// Create an empty string constant
+  ///
+  /// \param state: goto symex state
+  /// \param symex_assign: object handling symbol assignments
+  /// \param f_l1: application of function ID_cprover_string_empty_string_func
+  ///   with l1 renaming applied
+  void constant_propagate_empty_string(
```

Comment:
Sorry, I forgot to ask for unit tests for this function and the one below (ideally also the others that are added to `goto_symext`)